Load relation catalog for each PloneSite

busykoala opened the pull request:
Close #6

Since we need access to the ZODB on `bin/instance check_links` an
instance is started and we can access a PloneSite of our choice.

As we found out, the command mustn't contain normal dashes.

Commit:
Iterate all Sites in Installation

busykoala commented:
@lukasgraf wäre es dir möglich, dir diesen Commit kurz anzusehen und kurz Rückmeldung zu geben, bevor ich fortfahre?

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -0,0 +1,35 @@
+    relation_catalog = getUtility(ICatalog)
+
+
+if __name__ == '__main__':
```
This is not needed any more since the `main()` function is invoked directly by the `zopectl.command`. In fact, it's even counterproductive, since it implies that the `ftw/linkchecker/command/checking_links.py` can be started as a stand-alone script, which it can't.

busykoala replied:
👍

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -0,0 +1,35 @@
+    app = makerequest(plone)
+    setRequest(app.REQUEST)
+
+    houptding = app.restrictedTraverse('/')
```
Uhm, what? No. Use sensible english variable names, like `app = `. But this isn't needed, since `app` is already set up just above on line 13.

busykoala replied:
👍

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -0,0 +1,35 @@
+    # later on we need to iterate through
+    # each page and send reports to different people
+
+    MINI_SITE = houptding.get('Plone')
```
Nope. Use sensible english identifiers. Like `portal = ` (uppercase is used for constants by convention). See https://devdocs.4teamwork.ch/programmierrichtlinien/#einhalten-der-pep8-konventionen

busykoala replied:
👍

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -0,0 +1,35 @@
+from zope.globalrequest import setRequest
+import AccessControl
+from zc.relation.interfaces import ICatalog
+from zope.component import getUtility
```
Sort imports please. See https://devdocs.4teamwork.ch/programmierrichtlinien/#imports

busykoala replied:
👍

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -7,28 +7,47 @@
+    return getUtility(ICatalog)
+
+
+def suech_sache_im_katalog(relation_catalog):
```
WTF? https://devdocs.4teamwork.ch/programmierrichtlinien/#einhalten-der-pep8-konventionen

busykoala replied:
Thanks a lot for your review. I think I'm closer to the conventions now.

I was rather thinking of a review concerning the content and hope to get some advise on that too ;)

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -0,0 +1,51 @@
+    # make fake request to be able to traverse
+    app = makerequest(app)
+    setRequest(app.REQUEST)
+    overview_view = app.unrestrictedTraverse('plone-overview')
```
Is this a stock-Plone view? Didn't know this one before, but it seems to work.
But an alternative would be to simply list the objects inside the `app` (which is a container / folderish object):

(untested)
```python
from Products.CMFPlone.interfaces import IPloneSiteRoot

# ...
plone_sites = [obj for obj in app.objectValues() if IPloneSiteRoot.providedBy(obj)]
```
(which doesn't involve the use of a browser view which might change between Plone versions)

lukasgraf commented on ftw/linkchecker/command/checking_links.py:
```diff
@@ -0,0 +1,51 @@
+def get_relation_catalog_for_plone_site(app, PloneSiteId):
+    plone_site = app.get(PloneSiteId)
+
+    user = AccessControl.SecurityManagement.SpecialUsers.system
```
All the security manager related stuff should be moved out to the `main()` function, that's something you always want to run, and as early as possible. We usually create something like a `setup_plone()` helper function that does that, and returns the appropriately set up Plone site.

busykoala replied:
👍

lukasgraf commented on ftw/linkchecker/profiles/default/metadata.xml:
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0"?>
 <metadata>
     <dependencies>
+        <dependency>profile-plone.app.dexterity:default</dependency>
```
This might be unwanted - not sure if you really want to install Dexterity everywhere that the link-checker gets installed. But that should be reviewed by someone else, im not involved enough in these details of the project.

busykoala replied:
👍

lukasgraf reviewed (changes requested):
Just a few comments, otherwise looks good 👍 
Thanks for getting rid of the swiss german 😉

busykoala commented:
@lukasgraf I added the readme and removed the entry in `metadata`. The entry might originate from bobtemplates. But I tested without, so that shouldn't be a necessity.

Could you look over the changes again so I might be able to merge later on?
Thanks a lot :)

busykoala commented:
@lukasgraf could you check out the readme entry please?
@maethu ⏰

Commit:
Merge pull request #10 from 4teamwork/mo/mail_sending

Add a report mailer
